### PR TITLE
Update the explainer to reflect the new API

### LINKED ISSUE #24: Update documentation to align with a new API shape
See #15 for discussion.

Basically, the change is from having a lock object that is explicitly committed to an "acquireDisplayLock" which runs a callback, possibly continuing the scheduling via a context, and automatically commits.

### PR DESCRIPTION
**Pull request by vmpstr**
This closes #24 

Please take a look.

### CODE REVIEW

**Comment by vmpstr** [comment]
Also please use "squash and merge" since I seem to just have a bunch of commits in my attempts to rebase everything

**Comment by vmpstr** [comment]
@domenic, fyi

File: README.md [hunk 1]
```diff
@@ -32,40 +32,34 @@ locking an element.
...
-making the locked subtree `visibility: hidden`, meaning that we don't display
-any content while work is completed.
-
-There are other possible options here, such as displaying developer specified
-content. The details of this are up for discussion, but the consistent result is
-that the visual representation of the modified DOM is not necessarily rendered
-to screen, affording the user-agent the opportunity to break the work up into
-smaller pieces.
+The visual content of an element's subtree which is locked for display does not
+change. Specifically, if the element already existed in the DOM, then the
```
Comment (chrishtr) [inline README.md]: Add a sentence saying explicitly that the root element is not locked in the case of an element already in the DOM,
so updates to it are not cached.

File: explainer.md [hunk 2]
```diff
@@ -58,6 +58,13 @@ cases and limitations.
 Consider the following example.
 
 ```html
+ <style>
+   #container {
+    width: 100px;
+    height: 100px;
+    contain: strict;
```
Comment (chrishtr) [inline explainer.md]: contain: contents?

File: explainer.md [hunk 3]
```diff
@@ -219,78 +226,159 @@ user.
...
 
 ##### DOM modifications in the locked state
 
-When the element is in the locked state, modifications to it or its subtree’s
-DOM, or modifications to style that affect it or its subtree perform the
-following steps:
+When the element is in the locked state, the callback can make changes to the
+element's subtree. Specifically changes to DOM, or style that affect its subtree
+update the DOM in such a way that script can inspect it immediately, **with the
+exception of style-inducing properties** (see
```
Comment (chrishtr) [inline explainer.md]: style and layout

Comment (chrishtr) [reply explainer.md]: Also note that style and layout inducing properties, when queried, will force style or layout synchronously,

File: explainer.md [hunk 4]
```diff
@@ -219,78 +226,159 @@ user.
...
+##### Extending the lifetime of the lock
+
+It is possible that the passed callback, which is run to completion, runs for a
+long enough time to actually cause jank: it may prevent other script from
+running. A way to alleviate this is to schedule a continuation of the work using
+a DisplayLockContext, which is given to the callback.
+
+##### DisplayLockContext.schedule()
+
+The `acquireDisplayLock()` callback may call `context.schedule()` passing it
```
Comment (chrishtr) [inline explainer.md]: comma after schedule()`

File: explainer.md [hunk 5]
```diff
@@ -219,78 +226,159 @@ user.
...
+another callback which will schedule to run on the next frame. This allows for
```
Comment (chrishtr) [inline explainer.md]: "which will run as a separate event loop task in the future"

We don't map these directly to rendering.

File: explainer.md [hunk 6]
```diff
@@ -219,78 +226,159 @@ user.
...
+the callback to yield to other script, as well as other update phases that may
```
Comment (chrishtr) [inline explainer.md]: What do you mean by "as well as other update phases that may affect other content on the page"?

File: explainer.md [hunk 7]
```diff
@@ -219,78 +226,159 @@ user.
...
+##### DisplayLockContext.suspend()
+
+DisplayLockContext also allows script to indefinitely suspend processing and
+prevent the visual updates to appear on screen using
+`DisplayLockContext.suspend()`. When invoked, the following happens:
+
+- The lock lifetime is extended indefinitely.
+- Any scheduled continuations or acquireDisplayLock callbacks that have not yet
+  run will not run while the context is suspended.
+- A handle to a release object is returned, which allows script to resume the
```
Comment (chrishtr) [inline explainer.md]: "A handle object is returned"

File: explainer.md [hunk 8]
```diff
@@ -425,35 +516,41 @@ possbilities are listed below:
...
-
-window.onload = function() {
-  window.requestAnimationFrame(presentContent);
-};
-</script>
+ <style>
+   #container {
+    width: 100px;
+    height: 100px;
+    contain: strict;
```
Comment (chrishtr) [inline explainer.md]: contain: contents?

File: explainer.md [hunk 9]
```diff
@@ -425,35 +516,41 @@ possbilities are listed below:
...
+ </script>
 ```
 
 Similar to the original example, on load we present the content in the
-`complicated_subtree`. However, this time the function is async and we await the
-surrounding container’s display lock acquire first. This causes the user-agent
-to lock the container for visual updates. Then, we modify the DOM by setting the
-`complicated_subtree` display property to `block`, and call `commit()` on the
-container lock. This sequence of commands causes the user-agent to
+`complicated_subtree`. However, we now use acquireDisplayLock callback to modify
```
Comment (chrishtr) [inline explainer.md]: use the acquireDisplayLock callback

File: explainer.md [hunk 10]
```diff
@@ -425,35 +516,41 @@ possbilities are listed below:
...
+the contents. This causes the user-agent to lock the container for visual
```
Comment (chrishtr) [inline explainer.md]: "lock the container's subtree"

File: explainer.md [hunk 11]
```diff
@@ -488,6 +586,48 @@ without jank.
...
+* Frame 2 also detects that no more things have been scheduled and processes the
+  update phases to completion.
+* End of Frame 2: the promise is resolved. Note that here, script cannot
+  re-acquire the lock, since that would prevent the visual updates from making
+  it to screen; it must do this in a rAf or a setTimeout callback instead.
+* Frame 3: script calls acquireDisplayLock again, which has to finish the
+  current update phases and stash the paint output (as in the first step).
+* Frame 4: the callback is run, etc.
+
+If this animation continues like this, there is always two frames between script
```
Comment (chrishtr) [inline explainer.md]: "there are always"

File: explainer.md [hunk 12]
```diff
@@ -488,6 +586,48 @@ without jank.
...
+successively changing the animated properties (Frames 2 and 4 are the ones that
+run the callback). This means that the best frame rate the script may achieve
+with this version of display locking is 30fps.
+
+Since this is a common use case, we need to allow for a display lock driven
+animation to run at 60fps instead.
+
```
Comment (chrishtr) [inline explainer.md]: Also add a paragraph about "async subtree" (mode 1) vs "async element + async subtree" (mode 2)

File: explainer.md [hunk 13]
```diff
@@ -595,6 +735,17 @@ from complete.
       what is required to get the queried information. Another possibility is to
       simply disallow such queries into the locked subtree, failing with an
       invalid number returned.
+* Find-in-page and tab order.
+    * It is unclear whether the content put into a locked subtree should be
+      discoverable by find-in-page or tab order navigation. It is possible that
+      this decision will be left up to the script author by allowing an options
+      dictionary to be passed to acquireDisplayLock, which would dictate whether
+      the content modified is searcabhe (e.g. { searchable: true }).
```
Comment (chrishtr) [inline explainer.md]: searchable

File: explainer.md [hunk 14]
```diff
@@ -595,6 +735,17 @@ from complete.
...
+* Dealing with exceptions in callbacks.
+    * If a callback throws an exception, then we need to someone propagate that
```
Comment (chrishtr) [inline explainer.md]: s/someone/somehow/

File: explainer.md (hunk 12 above)
Comment (vmpstr) [reply explainer.md]: Note sure what you mean here? In what context do I bring up this difference? Or do you mean like a separate section?

File: explainer.md (hunk 6 above)
Comment (vmpstr) [reply explainer.md]: I've reworded this. I meant to say to update other parts of the page (the ones that are not locked)

**Comment by vmpstr** [comment]
PTAL. Updated everything with the exception of one comment above.

File: explainer.md (hunk 12 above)
Comment (chrishtr) [reply explainer.md]: In a separate example/paragraph..

**Comment by vmpstr** [comment]
PTAL, I added a new section about the different modes.

File: explainer.md [hunk 15]
```diff
@@ -488,6 +586,79 @@ without jank.
...
+### Locked subtree vs locked element + subtree
+
+Display locking supports two modes of operation, distinguished by when the lock
+is acquired
+
+#### Mode 1 (locked subtree)
+
+When the element's lock is acquired while the element is already a part of the
+DOM, then the draw commands associated with that element's subtree are stashed
+and used while the hold is acquired. Note that in this mode, changes to the
```
Comment (chrishtr) [inline explainer.md]: s/hold/lock/

File: explainer.md [hunk 16]
```diff
@@ -488,6 +586,79 @@ without jank.
...
+element itself (e.g. border, size) are updated synchronously. In other words,
+the element itself is not locked for dipslay, only its subtree.
```
Comment (chrishtr) [inline explainer.md]: nit: "display"

**Pull request merged**